src/test/custom_loader_test.cpp

```diff
@@ -0,0 +1,967 @@
+
+        cl_models_path = "/tmp/" + std::string(test_info->name());
+        cl_model_1_path = cl_models_path + "/model1/";
+        cl_model_2_path = cl_models_path + "/model2/";
+
+        const std::string FIRST_MODEL_NAME = "dummy";
+        const std::string SECOND_MODEL_NAME = "dummy_new";
+
+        std::filesystem::remove_all(cl_models_path);
+        std::filesystem::create_directories(cl_model_1_path);
+    }
+    void TearDown() {
+        // Clean up temporary destination
+        std::filesystem::remove_all(cl_models_path);
+    }
+    /**
+     * @brief This function should mimic most closely predict request to check for thread safety
+     */
+    void performPredict(const std::string modelName,
+        const ovms::model_version_t modelVersion,
+        const tensorflow::serving::PredictRequest& request,
+        std::unique_ptr<std::future<void>> waitBeforeGettingModelInstance = nullptr,
+        std::unique_ptr<std::future<void>> waitBeforePerformInference = nullptr);
+
+    void deserialize(const std::vector<float>& input, InferenceEngine::InferRequest& inferRequest, std::shared_ptr<ovms::ModelInstance> modelInstance) {
+        auto blob = InferenceEngine::make_shared_blob<float>(
+            modelInstance->getInputsInfo().at(DUMMY_MODEL_INPUT_NAME)->getTensorDesc(),
+            const_cast<float*>(reinterpret_cast<const float*>(input.data())));
+        inferRequest.SetBlob(DUMMY_MODEL_INPUT_NAME, blob);
+    }
+
+    void serializeAndCheck(int outputSize, InferenceEngine::InferRequest& inferRequest) {
+        std::vector<float> output(outputSize);
+        ASSERT_THAT(output, Each(Eq(0.)));
+        auto blobOutput = inferRequest.GetBlob(DUMMY_MODEL_OUTPUT_NAME);
+        ASSERT_EQ(blobOutput->byteSize(), outputSize * sizeof(float));
+        std::memcpy(output.data(), blobOutput->cbuffer(), outputSize * sizeof(float));
+        EXPECT_THAT(output, Each(Eq(2.)));
+    }
+
+    ovms::Status performInferenceWithRequest(const tensorflow::serving::PredictRequest& request, tensorflow::serving::PredictResponse& response) {
+        std::shared_ptr<ovms::ModelInstance> model;
+        std::unique_ptr<ovms::ModelInstanceUnloadGuard> unload_guard;
+        auto status = ovms::getModelInstance(manager, "dummy", 0, model, unload_guard);
+        if (!status.ok()) {
+            return status;
+        }
+
+        response.Clear();
+        return ovms::inference(*model, &request, &response, unload_guard);
+    }
+
+public:
+    ConstructorEnabledModelManager manager;
+    ovms::ModelConfig config = DUMMY_MODEL_CONFIG;
+    ~TestCustomLoader() {
+        std::cout << "Destructor of TestCustomLoader()" << std::endl;
+    }
+
+    std::string cl_models_path;
+    std::string cl_model_1_path;
+    std::string cl_model_2_path;
+};
+
+::grpc::Status test_PerformModelStatusRequestX(ModelServiceImpl& s, tensorflow::serving::GetModelStatusRequest& req, tensorflow::serving::GetModelStatusResponse& res) {

```

Comment:
What does the `X` stands for in `test_PerformModelStatusRequestX` and later when it's used with variables like `sx, reqx, resx`?

Reply:
Added a new function. Will have a relook at all the functions/part of code reused from the existing tests